cmd/minishift/cmd/start_preflight.go

```diff
@@ -0,0 +1,278 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/docker/machine/libmachine/drivers"
+	startFlags "github.com/minishift/minishift/cmd/minishift/cmd/config"
+	miniutil "github.com/minishift/minishift/pkg/minishift/util"
+
+	"github.com/minishift/minishift/pkg/util/os/atexit"
+	"github.com/spf13/viper"
+)
+
+const (
+	StorageDisk = "/mnt/sda1"
+)
+
+type preflightCheckRunner func() bool
+type preflightCheckWithDriverRunner func(driver drivers.Driver) bool
+
+func preflightCheckSucceedsOrFails(configNameOverrideIfSkipped string, execute preflightCheckRunner, message string, treatAsWarning bool, configNameOverrideIfWarning string, errorMessage string) {
+	fmt.Printf("-- %s ... ", message)
+
+	isSkip := viper.GetBool(configNameOverrideIfSkipped)
+	isWarn := viper.GetBool(configNameOverrideIfWarning)
+
+	if isSkip {
+		fmt.Println("SKIP")
+		return
+	}
+
+	if execute() {
+		fmt.Println("OK")
+		return
+	}
+
+	fmt.Println("FAIL")
+	errorMessage = fmt.Sprintf("   %s", errorMessage)
+	if isWarn {
+		fmt.Println(errorMessage)
+	} else {
+		atexit.ExitWithMessage(1, errorMessage)
+	}
+}
+
+func preflightCheckSucceedsOrFailsWithDriver(configNameOverrideIfSkipped string, execute preflightCheckWithDriverRunner, driver drivers.Driver, message string, treatAsWarning bool, configNameOverrideIfWarning string, errorMessage string) {
+	fmt.Printf("-- %s ... ", message)
+
+	isSkip := viper.GetBool(configNameOverrideIfSkipped)
+	isWarn := viper.GetBool(configNameOverrideIfWarning)
+
+	if isSkip {
+		fmt.Println("SKIP")
+		return
+	}
+
+	if execute(driver) {
+		fmt.Println("OK")
+		return
+	}
+
+	fmt.Println("FAIL")
+	errorMessage = fmt.Sprintf("   %s", errorMessage)
+	if isWarn || treatAsWarning {
+		fmt.Println(errorMessage)
+	} else {
+		atexit.ExitWithMessage(1, errorMessage)
+	}
+}
+
+func preflightChecksBeforeStartHost(vmDriver string) {
+	switch vmDriver {
+	case "xhyve":
+		preflightCheckSucceedsOrFails(
+			startFlags.SkipCheckXHyveDriver.Name,
+			checkXhyveDriver,
+			"Checking if 'xhyve' driver installed",
+			false, startFlags.WarnCheckXHyveDriver.Name,
+			"See install guide")
+	case "kvm":
+		preflightCheckSucceedsOrFails(
+			startFlags.SkipCheckKVMDriver.Name,
+			checkKvmDriver,
+			"Checking if 'kvm' driver installed",
+			false, startFlags.WarnCheckXHyveDriver.Name,
+			"See install guide")
+	case "hyperv":
+		preflightCheckSucceedsOrFails(
+			startFlags.SkipCheckHyperVDriver.Name,
+			checkHypervDriver,
+			"Checking if 'hyperv' driver configured",
+			false, startFlags.WarnCheckHyperVDriver.Name,
+			"Hyper-V virtual switch not set")
+	}
+}
+
+func checkXhyveDriver() bool {
+	_, err := exec.LookPath("docker-machine-driver-xhyve")
```

Comment:
>  I have also tried an alternative library

Which one? It did not work? I feel this really falls a bit short here, especially since it might make things even worse for the user in the case the driver exists and our check "passes", but the boot still fails.

I see two options.

* Add some additional message informing the user that the binary is there, but we are unable to verify the suid. If problems occur check the suid flag. (exact wording needs to be worked out)
* Use os.Exec to execute the driver, then use the returned Process to get the uid of the process and ensure it is '0'. Executing the binary might be a good thing regardless. since it also eliminates issues of the binary being corrupt. 

> but it already cost too much time so I moved on

I think we should get this right, especially since we put so much time into it.